### PR TITLE
refactor: Use the new Typesafe routes for navigation

### PR DESCRIPTION
**Pull request by soulcramer**
## 📋 Changes

Replace all the string routes by typed ones.
Encapsulate the navigation to be only accesible from specific destinations/graphs

## ✅ Checklist

- [x] I have reviewed the submitted code.
- [x] I have tested on a phone device/emulator.

### CODE REVIEW

File: catalog/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -43,6 +43,15 @@
...
             </intent-filter>
+            <intent-filter>
+                <action android:name="android.intent.action.VIEW" />
+
+                <category android:name="android.intent.category.DEFAULT" />
+                <category android:name="android.intent.category.BROWSABLE" />
+
+                <data android:scheme="https" />
+                <data android:host="android.spark.adevinta.com" />
+            </intent-filter>
```
Comment (SimonMarquis) [inline catalog/src/main/AndroidManifest.xml:54]: I don't think this was added on purpose 🤔

File: catalog/src/main/kotlin/com/adevinta/spark/catalog/configurator/Navigation.kt [hunk 2]
```diff
@@ -60,5 +58,6 @@ internal fun NavGraphBuilder.navGraph(
     }
 }
 
-internal const val ConfiguratorRoute = "configurator"
-internal const val ComponentIdArgName = "componentId"
+private fun NavController.navigateToConfiguratorShowcase(id: Int) {
+    navigate(route = ConfiguratorShowcase(componentId = id))
+}
```
Comment (SimonMarquis) [inline catalog/src/main/kotlin/com/adevinta/spark/catalog/configurator/Navigation.kt:63]: Do we need to extract this single usage into a function?
In other parts of the code, you don't extract them.
We should choose one consistent way of doing it. Personnaly, I would inline these at the call site directly.

File: catalog/src/main/kotlin/com/adevinta/spark/catalog/examples/Navigation.kt [hunk 3]
```diff
@@ -22,69 +22,67 @@
...
-internal const val ExampleRoute = "example"
-internal const val ComponentIdArgName = "componentId"
-internal const val ExampleIndexArgName = "exampleIndex"
+private fun NavController.navigateToExampleSection(id: Int) {
+    navigate(route = ExampleSection(componentId = id))
+}
+
+private fun NavController.navigateToExampleShowcase(id: Int, index: Int) {
+    navigate(route = ExampleShowcase(componentId = id, exampleIndex = index))
+}
```
Comment (SimonMarquis) [inline catalog/src/main/kotlin/com/adevinta/spark/catalog/examples/Navigation.kt:88]: Same here

File: catalog/src/main/AndroidManifest.xml (hunk 1 above)
Comment (soulcramer) [reply catalog/src/main/AndroidManifest.xml:54]: Ah actually it was but not directly but for a follow-up PR to add deep link support.
I'll readd it later

File: catalog/src/main/kotlin/com/adevinta/spark/catalog/examples/Navigation.kt (hunk 3 above)
Comment (soulcramer) [reply catalog/src/main/kotlin/com/adevinta/spark/catalog/examples/Navigation.kt]: Fixed in https://github.com/adevinta/spark-android/pull/1401/commits/b231b1cc07ea384f7c81c9a628d7d93e900eb9a1

File: catalog/src/main/kotlin/com/adevinta/spark/catalog/configurator/Navigation.kt (hunk 2 above)
Comment (soulcramer) [reply catalog/src/main/kotlin/com/adevinta/spark/catalog/configurator/Navigation.kt]: Fixed in https://github.com/adevinta/spark-android/pull/1401/commits/b231b1cc07ea384f7c81c9a628d7d93e900eb9a1

**Pull request merged**